rfcs/proposed/RFC-2-server-sdk-ncco-builders.md

```diff
@@ -0,0 +1,166 @@
+
+## Duration
+
+## Current State
+
+Proposed
+
+## Proposers
+
+Ben Greenberg
+
+## Detail
+
+The composition of valid NCCO actions are a required development component in constructing voice communications with the Vonage Voice API. The fine attention to idiosyncratic details that is required to create a correct NCCO is often a stumbling block for customers, both those who are new and those who are experienced. Vonage currently offers an [NCCO Reference Guide](https://developer.vonage.com/voice/voice-api/ncco-reference) on the API Developer Portal, but that guide does not provide real-time support for users developing in their environment. Additionally, the Vonage Voice API itself provides minimal error response support for invalid NCCO actions.
+
+The challenge of composing correct NCCO actions is one of the most frequently raised support issues in the various Vonage support challenges (Slack, Stack Overflow, GitHub Issues, etc.). All of our mainline SDKs fully support the Voice API, and provide to differing levels, support for the composition of NCCOs. This NCCO support though varies widely from one SDK to the next. In some of the SDKs the NCCO actions are treated as pass-through parameters with no validation or assistance, whereas in other SDKs there are more complete validators and builder support mechanisms. 
+
+We can offer a comprehensive NCCO builder in each of the mainline Vonage SDKs that will ease the NCCO action composition burden on our users considerably. These builders will assist users in both creating their NCCO actions, and provide real-time feedback on incorrect construction. 
+
+## Proposed Solution
+
+### Technical Specifications
+
+The NCCO builder MUST be instantiable. By providing distinct instantiations for each NCCO action, it becomes easier to provide specific feedback on each unique action for the developer. For example, there may be multiple `type` parameters throughout the NCCO actions, but the specifications for `type` will be different for each unique NCCO action. The disaggregation and unique instantiation of each action as its own class allows for that fine attention to detail.
+
+```ruby
+# Standalone NCCO with two actions
+
+talk = Vonage::Voice::Ncco.talk(text: 'Hello World!')
+input = Vonage::Voice::Ncco.input(type: ['dtmf'], dtmf: { bargeIn: true })
+ncco = Vonage::Voice::Ncco.build(talk, input)
+
+# => [{:action=>"talk", :text=>"Hello World!"}, {:action=>"input", :type=>["dtmf"], :dtmf=>{:bargeIn=>true}}]
+```
+
+Each NCCO action MUST be able to be accessed via its own method name accessed on the NCCO class. For example, the `connect` action would be accessed as `Vonage::Voice::Ncco.connect` and the `notify` action would be accessed as `Vonage::Voice::Ncco.notify`. For example, the following would accomplish this in the Ruby SDK:
```

Comment:
I would switch this from `MUST` to `MAY`. This reduces coupling between the builder class and the action classes, but still allows this structure as an option. It also helps sidestep the flexibility some languages provide where others don't or can't. 

```suggestion
Each NCCO action MAY be able to be accessed via its own method name accessed on the NCCO builder class. For example, the `connect` action would be accessed as `Vonage::Voice::Ncco.connect` and the `notify` action would be accessed as `Vonage::Voice::Ncco.notify`. For example, the following would accomplish this in the Ruby SDK:
```